contracts/staking/ServiceStakingToken.sol

```diff
@@ -0,0 +1,199 @@
+        bool success;
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+        // We'll write our calldata to this slot below, but restore it later.
+            let memPointer := mload(0x40)
+
+        // Write the abi-encoded calldata into memory, beginning with the function selector.
+            mstore(0, 0xa9059cbb00000000000000000000000000000000000000000000000000000000)
+            mstore(4, to) // Append the "to" argument.
+            mstore(36, amount) // Append the "amount" argument.
+
+            success := and(
+            // Set success to whether the call reverted, if not we check it either
+            // returned exactly 1 (can't just be non-zero data), or had no return data.
+                or(and(eq(mload(0), 1), gt(returndatasize(), 31)), iszero(returndatasize())),
+            // We use 68 because that's the total length of our calldata (4 + 32 * 2)
+            // Counterintuitively, this call() must be positioned after the or() in the
+            // surrounding and() because and() evaluates its arguments from right to left.
+                call(gas(), token, 0, 0, 68, 0, 32)
+            )
+
+            mstore(0x60, 0) // Restore the zero slot to zero.
+            mstore(0x40, memPointer) // Restore the memPointer.
+        }
+
+        if (!success) {
+            revert TransferFailed(token, address(this), to, amount);
+        }
+    }
+
+    /// @dev Checks token security deposit.
+    /// @param serviceId Service Id.
+    function _checkTokenSecurityDeposit(uint256 serviceId) internal view override {
+        // Get the service security token and deposit
+        (address token, uint96 securityDeposit) =
+            IServiceTokenUtility(serviceRegistryTokenUtility).mapServiceIdTokenDeposit(serviceId);
+
+        // The security token must match the contract token
+        if (securityToken != token) {
+            revert();
+        }
+
+        // The security deposit must be greater or equal to the minimum defined one
+        if (securityDeposit < minSecurityDeposit) {
+            revert();
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal override {
+        safeTransfer(securityToken, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    /// @dev Deposits funds for staking.
+    /// @param amount Token amount to deposit.
+    function deposit(uint256 amount) external {
+        // Distribute current staking rewards
+        _checkpoint(0);
+
+        // Add to the overall balance
+        safeTransferFrom(securityToken, msg.sender, address(this), amount);
+
+        // Add to the overall balance
+        uint256 newBalance = balance + amount;
+
+        // Update rewards per second
+        uint256 newRewardsPerSecond = (newBalance * apy) / (100 * 365 days);
```

Comment:
This strikes me as odd.

Reply:
Don't like how coupled transferring balance in is with checkpointing logic.

Reply:
The balance is added, the reward per second changes. I don't see the other way of doing that. Still think we can't mix different balance deposit over different times, as in my example above.

Bottom line - we need a checkpoint on deposits and on unstakes.

Reply:
no reward per second should be a constant
I see now what you misunderstood from the spec.

Nonces should purely be used for liveness check! It's a binary condition. Are enough tx made in a given time window? If yes, you get rewards as per "rewardsPerSecond" specifier. If not, no rewards.

Reply:
I believe this is exactly what's implemented `stakingRatio` or the `livenessCheck` is the amount of multisig tx-s made in a specific timeframe. If it is above the defined value, the service is eligible for reward, otherwise it's not. Rewards do not depend on the ratio, just a yes or no.

The available reward balance is then split between eligible services (satisfying the liveness check) from the time of the last checkpoint until a current block timestamp.